doc/barclamp.md

```diff
@@ -0,0 +1,1177 @@
+function Fpatch()
+  {
+  cd ~/crowbar-openstack
+  git diff "$COMMIT^" HEAD | patch --merge -p1 -d /opt/dell
+  for f in /opt/dell/*.yml
+    do
+    ln -s $f /opt/dell/crowbar_framework/barclamps 2>&1 > /dev/null
+    done
+  chmod 755 /opt/dell/bin/*
+  }
+
+sync_crowbar()
+  {
+  barclamp_install.rb /opt/dell/crowbar_framework/barclamps/ &&
+  for barclamp in $BARCLAMPS
+    do
+    knife cookbook upload -o /opt/dell/chef/cookbooks/ mybarclamp
+    done
+  systemctl restart crowbar
+  }
+```
+
+We will make heavy use of the `Cpatch`, `Fpatch` and `sync_crowbar` commands
+thus defined in the following sections. A short explanation of the three:
+
+* `Fpatch` will apply your whole fork against `/opt/dell` using `patch(1)`. You
+  use this command to apply the current state of your fork to a freshly
+  installed Crowbar node. This command may cause merge conflicts. If this
+  happens, resolve them before doing anything else.
+
+* `Cpatch` will apply the latest commit in `/root/crowbar-openstack` to
+  `/opt/dell` (use this after each new commit). This command may cause merge
+  conflicts. If this happens, resolve them before doing anything else.
+
+* `sync-crowbar` will make any updated state in `/opt/dell` appear in the
+  running Crowbar instance. Run this after every change, no matter how minor,
+  in `/opt/dell`.
+
+Using patches against `/opt/dell` may seem a bit over-the-top at first since
+one could simply copy files back and forth, but it prevents breakage in cases
+where `/opt/dell` is out of sync with the `crowbar-openstack` repository's
+master branch or your own fork of `crowbar-openstack` is out of sync with
+upstream: in the former case copying files from your fork to `/opt/dell` may
+cause an inconsistent state and errors entirely unrelated to your change. In
+the latter case copying files from `/opt/dell` to your fork may re-introduce
+old code that is no longer present upstream. Both should not be an issue for
+new barclamps, but it is better to get used to the `patch(1)` based workflow
+right away since these problems do become a problem when changing an existing
+barclamp later and thus the `patch(1)` based workflow becomes neccessary.
+
+## Barclamp Component Overview
+
+A barclamp consists of two main components:
+
+* A [chef cookbook](https://docs.chef.io/cookbooks.html) which takes care of
+  actually deploying your service.
+* Some glue code to tie your chef cookbook into the Crowbar Web UI. This glue
+  code will parametrize your chef cookbook and apply its component roles to the
+  nodes it orchestrates.
+
+The subsequent sections will describe these two components and the ingredients
+that go into them into detail. They start of with an overview of a barclamp's
+development phases as we know them from experience and take you through all
+aspects of barclamp implementation in the order you are likely to encounter
+them. They are meant to be read in parallel to developing your own barclamp.
+
+## Development Phases
```

Comment:
Too late - some of the necessary ingredients already got deleted: https://github.com/crowbar/crowbar-core/pull/1026#issuecomment-275105884